# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/angle/checkout/src/libANGLE/renderer/Format.h
# Commit: 3a91885d47e0
# Full Hash: 3a91885d47e0ef7fba9749562d52300eea67c651
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2019-11-11 21:52:52
# Regressor Bug: 1578576
# File Overlap Count: 4
# Description:
#   Bug 1578576 - Part 2: Update to ANGLE 3865 r=jgilbert
#   
#   Depends on D44561
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D44579
# ==============================================================================

diff -r 77cee77c0307 -r 3a91885d47e0 gfx/angle/checkout/src/libANGLE/renderer/Format.h
--- a/gfx/angle/checkout/src/libANGLE/renderer/Format.h	Fri Nov 08 16:41:42 2019 +0000
+++ b/gfx/angle/checkout/src/libANGLE/renderer/Format.h	Fri Nov 08 17:06:00 2019 +0000
@@ -41,7 +41,9 @@
                             GLuint pixelBytes,
                             GLuint componentAlignmentMask,
                             bool isBlock,
-                            bool isFixed);
+                            bool isFixed,
+                            bool isScaled,
+                            gl::VertexAttribType vertexAttribType);
 
     static const Format &Get(FormatID id) { return gFormatInfoTable[static_cast<int>(id)]; }
 
@@ -49,14 +51,17 @@
 
     constexpr bool hasDepthOrStencilBits() const;
     constexpr bool isLUMA() const;
-    constexpr GLuint channelCount() const;
 
-    constexpr bool isInt() const;
+    constexpr bool isSint() const;
     constexpr bool isUint() const;
     constexpr bool isSnorm() const;
     constexpr bool isUnorm() const;
     constexpr bool isFloat() const;
 
+    constexpr bool isInt() const { return isSint() || isUint(); }
+    constexpr bool isNorm() const { return isSnorm() || isUnorm(); }
+    constexpr bool isPureInt() const { return isInt() && !isScaled; }
+
     bool operator==(const Format &other) const { return this->id == other.id; }
 
     FormatID id;
@@ -93,10 +98,29 @@
     // 0x0.
     GLuint componentAlignmentMask;
 
+    GLuint channelCount;
+
     bool isBlock;
     bool isFixed;
+    bool isScaled;
+
+    // For vertex formats only. Returns the "type" value for glVertexAttribPointer etc.
+    gl::VertexAttribType vertexAttribType;
 };
 
+constexpr GLuint GetChannelCount(GLuint redBits,
+                                 GLuint greenBits,
+                                 GLuint blueBits,
+                                 GLuint alphaBits,
+                                 GLuint luminanceBits,
+                                 GLuint depthBits,
+                                 GLuint stencilBits)
+{
+    return (redBits > 0 ? 1 : 0) + (greenBits > 0 ? 1 : 0) + (blueBits > 0 ? 1 : 0) +
+           (alphaBits > 0 ? 1 : 0) + (luminanceBits > 0 ? 1 : 0) + (depthBits > 0 ? 1 : 0) +
+           (stencilBits > 0 ? 1 : 0);
+}
+
 constexpr Format::Format(FormatID id,
                          GLenum glFormat,
                          GLenum fboFormat,
@@ -115,7 +139,9 @@
                          GLuint pixelBytes,
                          GLuint componentAlignmentMask,
                          bool isBlock,
-                         bool isFixed)
+                         bool isFixed,
+                         bool isScaled,
+                         gl::VertexAttribType vertexAttribType)
     : id(id),
       glInternalFormat(glFormat),
       fboImplementationInternalFormat(fboFormat),
@@ -133,8 +159,17 @@
       stencilBits(stencilBits),
       pixelBytes(pixelBytes),
       componentAlignmentMask(componentAlignmentMask),
+      channelCount(GetChannelCount(redBits,
+                                   greenBits,
+                                   blueBits,
+                                   alphaBits,
+                                   luminanceBits,
+                                   depthBits,
+                                   stencilBits)),
       isBlock(isBlock),
-      isFixed(isFixed)
+      isFixed(isFixed),
+      isScaled(isScaled),
+      vertexAttribType(vertexAttribType)
 {}
 
 constexpr bool Format::hasDepthOrStencilBits() const
@@ -149,13 +184,7 @@
     return redBits == 0 && (luminanceBits > 0 || alphaBits > 0);
 }
 
-constexpr GLuint Format::channelCount() const
-{
-    return (redBits > 0) + (greenBits > 0) + (blueBits > 0) + (alphaBits > 0) +
-           (luminanceBits > 0) + (depthBits > 0) + (stencilBits > 0);
-}
-
-constexpr bool Format::isInt() const
+constexpr bool Format::isSint() const
 {
     return componentType == GL_INT;
 }